File: Development-Data-Process/src/feature_extractor.py
import re
import ipaddress
import requests
from bs4 import BeautifulSoup
from urllib.parse import urlparse
import tldextract

# --- 1. URL-Based Features (11 Features) ---

def having_IP(url):
    try:
        domain = urlparse(url).netloc
        ipaddress.ip_address(domain)
        return 1
    except ValueError:
        return 0

def URL_Length(url):
    return 1 if len(url) > 75 else 0

def Shortining_Service(url):
    shortening_services = [
        "bit.ly", "goo.gl", "shorte.st", "go2l.ink", "x.co", "ow.ly", "t.co", "tinyurl", "tr.im", "is.gd", "cli.gs", "yfrog.com", "migre.me", "ff.im", "tiny.cc", "url4.eu", "twit.ac", "su.pr", "twurl.nl", "snipurl.com", "short.to", "budurl.com", "ping.fm", "post.ly", "just.as", "bkite.com", "snipr.com", "fic.kr", "loopt.us", "doiop.com", "short.ie", "kl.am", "wp.me", "rubyurl.com", "om.ly", "to.ly", "bit.do", "t.co", "lnkd.in", "db.tt", "qr.ae", "adf.ly", "goo.gl", "bitly.com", "cur.lv", "tinyurl.com", "ow.ly", "bit.ly", "ity.im", "q.gs", "is.gd", "po.st", "bc.vc", "twitthis.com", "u.to", "j.mp", "buzurl.com", "cutt.us", "u.bb", "yourls.org", "x.co", "prettylinkpro.com", "scrnch.me", "filoops.info", "vzturl.com", "qr.net", "1url.com", "tweez.me", "v.gd", "tr.im", "link.zip.net"
    ]
    domain = urlparse(url).netloc
    for service in shortening_services:
        if service in domain:
            return 1
    return 0

def having_At_Symbol(url):
    return 1 if "@" in url else 0

def double_slash_redirecting(url):
    return 1 if url.rfind('//') > 7 else 0

def Prefix_Suffix(url):
    domain = urlparse(url).netloc
    return 1 if "-" in domain else 0

def having_Sub_Domain(url):
    ext = tldextract.extract(url)
    subdomain = ext.subdomain
    if not subdomain: return 0
    if subdomain == 'www': return 0
    return 1 if subdomain.count('.') > 0 else 0

# --- NEW HIGH-DEF FEATURES ---
def Hyphen_Count(url):
    domain = urlparse(url).netloc
    return domain.count('-')

def Dot_Count(url):
    domain = urlparse(url).netloc
    if domain.startswith('www.'):
        domain = domain[4:]
    return domain.count('.')

def Numeric_Chars(url):
    domain = urlparse(url).netloc
    return sum(c.isdigit() for c in domain)

def Is_Common_TLD(url):
    # 0 = Safe/Common, 1 = Risky/Uncommon
    common_tlds = ['com', 'org', 'net', 'edu', 'gov', 'uk', 'us', 'ca', 'au', 'de', 'fr', 'it', 'jp', 'io', 'co', 'ai']
    ext = tldextract.extract(url)
    return 0 if ext.suffix in common_tlds else 1


# --- 2. HTML & Content-Based Features (5 Features) ---

def get_html_features(url):
    feature_dict = {
        'SFH': 0,
        'Submitting_to_email': 0,
        'on_mouseover': 0,
        'RightClick': 0,
        'Iframe': 0
    }
    
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'}
        response = requests.get(url, headers=headers, timeout=3)
        soup = BeautifulSoup(response.text, 'html.parser')
        html_lower = response.text.lower()

        # 1. SFH
        for form in soup.find_all('form'):
            action = form.get('action')
            if not action or action == "about:blank" or action == "":
                feature_dict['SFH'] = 1
                break

        # 2. Email
        for form in soup.find_all('form'):
            action = form.get('action')
            if action and "mailto:" in action:
                feature_dict['Submitting_to_email'] = 1
                break

        # 3. on_mouseover
        if "onmouseover" in html_lower:
            feature_dict['on_mouseover'] = 1

        # 4. RightClick
        if "event.button==2" in html_lower or "oncontextmenu" in html_lower:
            feature_dict['RightClick'] = 1

        # 5. Iframe
        if len(soup.find_all('iframe')) > 0:
            feature_dict['Iframe'] = 1

    except:
        pass 
    
    return feature_dict

# --- 3. Main Function (Exports 16 Features) ---
FINAL_FEATURE_LIST = [
    'having_IP', 'URL_Length', 'Shortining_Service', 'having_At_Symbol',
    'double_slash_redirecting', 'Prefix_Suffix', 'having_Sub_Domain',
    'Hyphen_Count', 'Dot_Count', 'Numeric_Chars', 'Is_Common_TLD',
    'SFH', 'Submitting_to_email', 'on_mouseover', 'RightClick', 'Iframe'
]

def extract_all_features(url):
    features = []
    
    # URL Features
    features.append(having_IP(url))
    features.append(URL_Length(url))
    features.append(Shortining_Service(url))
    features.append(having_At_Symbol(url))
    features.append(double_slash_redirecting(url))
    features.append(Prefix_Suffix(url))
    features.append(having_Sub_Domain(url))
    
    # New URL Features
    features.append(Hyphen_Count(url))
    features.append(Dot_Count(url))
    features.append(Numeric_Chars(url))
    features.append(Is_Common_TLD(url))

    # HTML Features
    html_f = get_html_features(url)
    features.append(html_f['SFH'])
    features.append(html_f['Submitting_to_email'])
    features.append(html_f['on_mouseover'])
    features.append(html_f['RightClick'])
    features.append(html_f['Iframe'])

    return features

File: PhishShield/feature_extractor.js
// Feature Extractor - 16 ROBUST FEATURES

function isIP(domain) {
    const ipRegex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
    return ipRegex.test(domain);
}

function extractFeaturesFromPage() {
    try {
        const url = window.location.href;
        const urlObj = new URL(url);
        let domain = urlObj.hostname;
        const features = [];

        // --- URL Features (7 Original) ---
        features.push(isIP(domain) ? 1 : 0); // 1
        features.push(url.length > 75 ? 1 : 0); // 2
        
        const shorteners = ["bit.ly", "goo.gl", "shorte.st", "go2l.ink", "x.co", "ow.ly", "t.co", "tinyurl", "tr.im", "is.gd", "cli.gs", "yfrog.com", "migre.me", "ff.im", "tiny.cc", "url4.eu", "twit.ac", "su.pr", "twurl.nl", "snipurl.com", "short.to", "budurl.com", "ping.fm", "post.ly", "just.as", "bkite.com", "snipr.com", "fic.kr", "loopt.us", "doiop.com", "short.ie", "kl.am", "wp.me", "rubyurl.com", "om.ly", "to.ly", "bit.do", "lnkd.in", "db.tt", "qr.ae", "adf.ly", "cur.lv", "ity.im", "q.gs", "po.st", "bc.vc", "twitthis.com", "u.to", "j.mp", "buzurl.com", "cutt.us", "u.bb", "yourls.org", "prettylinkpro.com", "scrnch.me", "filoops.info", "vzturl.com", "qr.net", "1url.com", "tweez.me", "v.gd", "link.zip.net"];
        features.push(shorteners.some(s => domain.includes(s)) ? 1 : 0); // 3

        features.push(url.includes("@") ? 1 : 0); // 4
        features.push(url.lastIndexOf("//") > 7 ? 1 : 0); // 5
        features.push(domain.includes("-") ? 1 : 0); // 6

        const parts = domain.split('.').filter(p => p !== 'www');
        features.push(parts.length > 3 ? 1 : 0); // 7

        // --- NEW FEATURES (4) ---
        
        // 8. Hyphen_Count
        features.push((domain.match(/-/g) || []).length);

        // 9. Dot_Count (Ignore www)
        let cleanDomain = domain.startsWith("www.") ? domain.substring(4) : domain;
        features.push((cleanDomain.match(/\./g) || []).length);

        // 10. Numeric_Chars
        features.push((domain.match(/\d/g) || []).length);

        // 11. Is_Common_TLD (0=Safe, 1=Risk)
        // We extract the suffix roughly (last part). 
        const suffix = parts[parts.length - 1];
        const common_tlds = ['com', 'org', 'net', 'edu', 'gov', 'uk', 'us', 'ca', 'au', 'de', 'fr', 'it', 'jp', 'io', 'co', 'ai'];
        features.push(common_tlds.includes(suffix) ? 0 : 1);


        // --- HTML Features (5 Original) --- 
        const forms = Array.from(document.querySelectorAll('form'));
        let sfh = 0;
        if (forms.some(f => !f.action || f.action === "about:blank" || f.action === "")) sfh = 1;
        features.push(sfh); // 12

        features.push(forms.some(f => f.action && f.action.startsWith('mailto:')) ? 1 : 0); // 13

        const htmlStr = document.documentElement.outerHTML.toLowerCase();
        features.push(htmlStr.includes('onmouseover') ? 1 : 0); // 14
        features.push((htmlStr.includes('event.button==2') || htmlStr.includes('oncontextmenu')) ? 1 : 0); // 15
        features.push(document.querySelectorAll('iframe').length > 0 ? 1 : 0); // 16

        console.log("PhishShield 16 Features:", features);
        return features;
    } catch (e) {
        console.error("Feature Extraction Error", e);
        return new Array(16).fill(0); 
    }
}
// --- NLP CONTENT SCANNER (Phase 2) ---

function calculateNLPScore() {
    try {
        const textContent = document.body.innerText.toLowerCase();
        let nlpScore = 0;

        // 1. Define Keyword Groups
        const urgencyWords = ['suspended', 'locked', 'banned', 'authorized', 'prevent', 'immediately', 'urgent', 'shutdown', 'deactivate'];
        const sensitiveWords = ['password', 'credit card', 'social security', 'bank account', 'routing number', 'pin', 'verify identity'];
        const actionWords = ['click here', 'login', 'sign in', 'update now', 'confirm', 'restore'];

        // 2. Scan for presence (Boolean)
        const hasUrgency = urgencyWords.some(w => textContent.includes(w));
        const hasSensitive = sensitiveWords.some(w => textContent.includes(w));
        const hasAction = actionWords.some(w => textContent.includes(w));

        // 3. Logic Scoring (Combinations are deadly)
        
        // Scenario A: "Your account is suspended" (Urgency + Sensitive)
        // Common in: Account lockout scams
        if (hasUrgency && hasSensitive) {
            console.log("NLP: Urgency + Sensitive keywords detected (+0.25)");
            nlpScore += 0.25;
        }

        // Scenario B: "Click here to login" (Action + Sensitive)
        // Common in: Credential harvesting
        if (hasAction && hasSensitive) {
            console.log("NLP: Action + Sensitive keywords detected (+0.15)");
            nlpScore += 0.15;
        }

        // Scenario C: "Urgent update required" (Urgency + Action)
        if (hasUrgency && hasAction) {
            console.log("NLP: Urgency + Action keywords detected (+0.15)");
            nlpScore += 0.15;
        }

        return nlpScore;

    } catch (e) {
        console.error("NLP Scan Failed", e);
        return 0;
    }
}